### PR TITLE
LLDP: Add packet validation

### PR DESCRIPTION
**Pull request by ssahani**
LLDP type system name and system description should
be with in 255 characters and unique.

Let's add the validation to discard corrupt packets.

### CODE REVIEW

File: src/libsystemd-network/sd-lldp.c [hunk 1]
```diff
@@ -301,11 +289,50 @@ int lldp_handle_packet(tlv_packet *tlv, uint16_t length) {
                         ttl = true;
 
                         break;
+                case LLDP_TYPE_SYSTEM_NAME:
+
+                        /* According to RFC 1035 the length of a FQDN is limited to 255 characters */
+                        if(len > 255) {
```
Comment (zonque) [inline src/libsystemd-network/sd-lldp.c]: Whitespace before `(`

File: src/libsystemd-network/sd-lldp.c [hunk 2]
```diff
@@ -301,11 +289,50 @@ int lldp_handle_packet(tlv_packet *tlv, uint16_t length) {
...
+                                goto out;
+                        }
+
+                        system_name = true;
+
+                        break;
+                case LLDP_TYPE_SYSTEM_DESCRIPTION:
+
+                        /* 0 <= n <= 255 octets */
+                        if(len > 255) {
```
Comment (zonque) [inline src/libsystemd-network/sd-lldp.c]: See above.

File: src/libsystemd-network/sd-lldp.c [hunk 3]
```diff
@@ -301,11 +289,50 @@ int lldp_handle_packet(tlv_packet *tlv, uint16_t length) {
...
+
+                        system_description = true;
+
+                        break;
                 default:
 
                         if (len == 0) {
-                                log_lldp("TLV type = %d's, length 0 received . Dropping ...",
-                                         type);
+                                log_lldp("TLV type = %d's, length 0 received. Dropping .", type);
```
Comment (zonque) [inline src/libsystemd-network/sd-lldp.c]: No whitespace before a full stop, please.

File: src/libsystemd-network/sd-lldp.c [hunk 4]
```diff
@@ -315,7 +342,7 @@ int lldp_handle_packet(tlv_packet *tlv, uint16_t length) {
         }
 
         if(!chassis_id || !port_id || !ttl || !end) {
-                log_lldp( "One or more mandotory TLV missing . Dropping ...");
+                log_lldp( "One or more mandotory TLV missing. Dropping .");
```
Comment (zonque) [inline src/libsystemd-network/sd-lldp.c]: See above.

File: src/libsystemd-network/sd-lldp.c [hunk 5]
```diff
@@ -324,7 +351,7 @@ int lldp_handle_packet(tlv_packet *tlv, uint16_t length) {
 
         r = tlv_packet_parse_pdu(tlv, length);
         if (r < 0) {
-                log_lldp( "Failed to parse the TLV. Dropping ...");
+                log_lldp( "Failed to parse the TLV. Dropping .");
```
Comment (zonque) [inline src/libsystemd-network/sd-lldp.c]: See above.

File: src/libsystemd-network/sd-lldp.c [hunk 6]
```diff
@@ -183,8 +179,7 @@ int lldp_handle_packet(tlv_packet *tlv, uint16_t length) {
 
                 if (type == LLDP_TYPE_END) {
                         if (len != 0) {
-                                log_lldp("TLV type end is not length 0. Length:%d received . Dropping ...",
-                                         len);
+                                log_lldp("TLV type end is not length 0. Length: '%d' received. Dropping .", len);
```
Comment (zonque) [inline src/libsystemd-network/sd-lldp.c]: No whitespace before a full stop. This also applies to many other locations below.

File: src/libsystemd-network/sd-lldp.c [hunk 7]
```diff
@@ -221,15 +215,14 @@ int lldp_handle_packet(tlv_packet *tlv, uint16_t length) {
...
-                                         len);
+                                log_lldp("Received malformed Chassis ID TLV len = '%d'. Dropping.", len);
 
                                 malformed = true;
                                 goto out;
                         }
 
                         if (chassis_id) {
-                                log_lldp("Duplicate Chassis ID TLV found. Dropping ...");
+                                log_lldp("Duplicate Chassis ID TLV found. Dropping.");
```
Comment (zonque) [inline src/libsystemd-network/sd-lldp.c:225]: Shouldn't we give a little more context to the person possibly reading this log? I don't know the code very well, so I can't say what details deserve logging, but as it stands, it doesn't help much I fear.

Comment (ssahani) [inline src/libsystemd-network/sd-lldp.c:225]: Hmm  . LLDP allows only one chassis ID field .  So we just discard it if we get two.

Comment (zonque) [inline src/libsystemd-network/sd-lldp.c:225]: That's not the point. The packet that causes this problem was sent from some other host, and so we should at least log that host's MAC or something. What action is an administrator supposed to take when this message appears in the logs? What benefit does it have in its current form?

Comment (ssahani) [inline src/libsystemd-network/sd-lldp.c:225]: That would be a separate patch .  And it's not only that it could be corrupted.
LLDP_TYPE_CHASSIS_ID contains 

```
      LLDP_CHASSIS_SUBTYPE_CHASSIS_COMPONENT   = 1,                                                                                                                                    
        LLDP_CHASSIS_SUBTYPE_INTERFACE_ALIAS     = 2,                                                                                                                                    
        LLDP_CHASSIS_SUBTYPE_PORT_COMPONENT      = 3,                                                                                                                                    
        LLDP_CHASSIS_SUBTYPE_MAC_ADDRESS         = 4,                                                                                                                                    
        LLDP_CHASSIS_SUBTYPE_NETWORK_ADDRESS     = 5,                                                                                                                                    
        LLDP_CHASSIS_SUBTYPE_INTERFACE_NAME      = 6,                                                                                                                                    
        LLDP_CHASSIS_SUBTYPE_LOCALLY_ASSIGNED    = 7,     
```

Just for logging I don't see decoding  all the subtypes and print.

Comment (zonque) [inline src/libsystemd-network/sd-lldp.c:225]: I don't get it. You have `tlv` and `port` as local variables, and both contain important information about the context of the packet. Why not log them if there's something wrong with the packet you're parsing? Again: what benefit for the administrator does the log message have as it currently stands?

Comment (ssahani) [inline src/libsystemd-network/sd-lldp.c:225]: Again No . These logs are debug logs .  You don't see them unless run in debug mode. We don't do usual log in libs at all.  I am not convinced at all

Comment (zonque) [inline src/libsystemd-network/sd-lldp.c:225]: And I am not convinced they are of any help during debugging if they don't give the admin an idea of what's really going on and which peer is sending bogus data.

Comment (ssahani) [inline src/libsystemd-network/sd-lldp.c:225]: How do you know which peer is sending wrong data ?  Again Does this PR says add debugging information ? This  PR is for domain validation.

File: src/libsystemd-network/sd-lldp.c [hunk 8]
```diff
@@ -315,7 +342,7 @@ int lldp_handle_packet(tlv_packet *tlv, uint16_t length) {
         }
 
         if(!chassis_id || !port_id || !ttl || !end) {
-                log_lldp( "One or more mandotory TLV missing . Dropping ...");
+                log_lldp("One or more mandotory TLV missing. Dropping.");
```
Comment (torstehu) [inline src/libsystemd-network/sd-lldp.c]: should be... more mandatory TLV..

Comment (ssahani) [inline src/libsystemd-network/sd-lldp.c]: yes right ! will fix it .
